src/main/java/org/kiwiproject/registry/consul/server/ConsulRegistryService.java

```diff
@@ -0,0 +1,220 @@
+
+    /**
+     * Maximum number of attempts we will make trying to register with Eureka (via POST)
+     */
+    @VisibleForTesting
+    static final int MAX_REGISTRATION_ATTEMPTS = 60;
+
+    /**
+     * Value for delay to wait if a call to Eureka fails.
+     */
+    private static final long RETRY_DELAY = 1;
+
+    /**
+     * Unit for delay to wait if a call to Eureka fails.
+     */
+    private static final TimeUnit RETRY_DELAY_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * Value for delay when un-registering if a call to Eureka fails.
+     */
+    private static final long UNREGISTER_RETRY_DELAY = 3;
+
+    /**
+     * Time unit for delay when un-registering if a call to Eureka fails.
+     */
+    private static final TimeUnit UNREGISTER_RETRY_DELAY_UNIT = TimeUnit.SECONDS;
+
+    /**
+     * Maximum number of attempts we will make to un-register from Eureka (via DELETE request).
+     */
+    private static final int MAX_UNREGISTER_ATTEMPTS = 5;
+
+    private final Consul consul;
+    private final ConsulRegistrationConfig config;
+    private final SimpleRetryer registerRetryer;
+    private final SimpleRetryer unregisterRetryer;
+
+    @VisibleForTesting
+    final AtomicReference<ServiceInstance> registeredService;
+
+    public ConsulRegistryService(Consul consul, ConsulRegistrationConfig config, KiwiEnvironment environment) {
+        this.consul = consul;
+        this.config = config;
+        this.registeredService = new AtomicReference<>();
+
+        this.registerRetryer = SimpleRetryer.builder()
+                .environment(environment)
+                .maxAttempts(MAX_REGISTRATION_ATTEMPTS)
+                .retryDelayTime(RETRY_DELAY)
+                .retryDelayUnit(RETRY_DELAY_UNIT)
+                .build();
+
+        this.unregisterRetryer = SimpleRetryer.builder()
+                .environment(environment)
+                .maxAttempts(MAX_UNREGISTER_ATTEMPTS)
+                .retryDelayTime(UNREGISTER_RETRY_DELAY)
+                .retryDelayUnit(UNREGISTER_RETRY_DELAY_UNIT)
+                .build();
+    }
+
+    @Override
+    public ServiceInstance createCandidateFrom(ServiceInfo serviceInfo) {
+        return ServiceInstance.fromServiceInfo(serviceInfo)
+                .withStatus(ServiceInstance.Status.UP);
+    }
+
+    @Override
+    public ServiceInstance register(ServiceInstance serviceToRegister) {
+        checkState(!isRegistered(), "Cannot register. Already managing a registered instance");
+
+        var registration = fromServiceInstance(serviceToRegister);
+
+        var service = registerRetryer.tryGetObject(() -> {
+            consul.agentClient().register(registration);
+            return serviceToRegister.withInstanceId(registration.getId());
+        });
+
+        if (service.isEmpty()) {
+            LOG.error("Registration failed for service {} with id {}. See logs above", serviceToRegister.getServiceName(), registration.getId());
```

Comment:
Is "See logs above" correct? I am guessing it means that some errors would have been logged already, but (obviously) not by _this_ method?

Reply:
Yes there would have been errors logged during the retry section above